merkle-tree/hasher/src/zero_bytes/poseidon.rs

```diff
@@ -3,168 +3,167 @@
 use super::ZeroBytes;
 pub const ZERO_BYTES: ZeroBytes = [
     [
-        32u8, 27u8, 47u8, 234u8, 113u8, 52u8, 186u8, 64u8, 78u8, 135u8, 6u8, 76u8, 82u8, 67u8,
-        144u8, 244u8, 144u8, 220u8, 214u8, 156u8, 168u8, 133u8, 18u8, 188u8, 227u8, 249u8, 224u8,
-        48u8, 227u8, 58u8, 66u8, 40u8,
+        0u8, 0u8, 0u8, 0u8, 0u8, 0u8, 0u8, 0u8, 0u8, 0u8, 0u8, 0u8, 0u8, 0u8, 0u8, 0u8, 0u8, 0u8,
+        0u8, 0u8, 0u8, 0u8, 0u8, 0u8, 0u8, 0u8, 0u8, 0u8, 0u8, 0u8, 0u8, 0u8,
```

Comment:
is this the leaf value or the leaf hash value

Reply:
It's the leaf value and also the first subtree value. The thing is that their implementation doesn't hash inserted leaves on the most bottom level. They assume that your leaf is something that's already hashed and they care about hashing parent nodes only.

It confused me a bit as well and that's different from many other implementations and from what Wikipedia says. Let's compare it in details.

## The previous model

The Wikipedia definition and what we are used to think about is like:

```
             H(...)
       /----/      \----\
    H(...)            H(...)
   /      \          /      \
H(L1)    H(L2)    H(L3)    H(L4)  (hashes of data blocks, the bottom row of the MT)
  |        |        |        |
--------------------------------
  |        |        |        |
 L1        L2      L3        L4   (data blocks, not a part of the MT)

```

In this model, it makes sense to think of zero bytes in the following way:

```
                  H(...)
          /-----/        \-----\
       H(...)                H(...)
     /        \            /        \
H([0;32])  H([0;32])  H([0;32])  H([0;32])  (hashes of data blocks, the bottom row of the MT)
    |          |          |          |
------------------------------------------
    |          |          |          |
  [0;32]     [0;32]     [0;32]     [0;32]   (data blocks, not a part of the MT)
```

That's what we are doing currently (before this PR), therefore the first zero byte is a non-zero hash already. In our mental model, we treat the bottom row of MT as set of hashed leaves.

## spl-concurrent-merkle-tree model

Their model says:

* the leaves you insert are just becoming the part of the Merkle tree
  * they are probably some hashes, but we don't care about that
  * we only care about creating hashes of parents

So it looks like:

```
1st subtree
|
|  2nd subtree  
|    |
v    v

          H(...)
         /      \
   H(...)        H(...)
  /      \      /      \
L1        L2  L3        L4  (this is already the part of the MT)
```

So in this model, zero bytes are:

```
1st subtree
  |
  |  2nd subtree  
  |   |
  v   v

            H(...)
          /        \
    H(...)          H(...)
   /      \        /      \
[0;32]  [0;32]  [0;32]  [0;32]  (this is already the part of the MT)
```

When you start inserting leaves, the only expectation for them is to be a 32-byte array. They immediately become the part of MT. Hashing is done on caller's side.

After the first insertion, the tree looks like

```
1st subtree
  |
  |  2nd subtree  
  |   |
  v   v

            H(...)
          /        \
    H(...)          H(...)
   /      \        /      \
 L1     [0;32]  [0;32]  [0;32]  (this is already the part of the MT)
```

You usually would hash L1 yourself, but, again, the MT doesn't care. You can insert anything which satisfies the `[u8; 32]` type and it's your responsibility to hash it.

In case of compressed NFTs, spl-concurrent-merkle-tree completely doesn't care about hashing the bottom leaves. It's bubblegum which hashes the NFT (with keccak) before insertion:

https://github.com/metaplex-foundation/mpl-bubblegum/blob/cea5f00a98778ca4e73694d7b418acd21483c0af/programs/bubblegum/program/src/processor/mint.rs#L153-L161

https://github.com/metaplex-foundation/mpl-bubblegum/blob/cea5f00a98778ca4e73694d7b418acd21483c0af/programs/bubblegum/program/src/state/leaf_schema.rs#L117-L138